code/drasil-lang/lib/Language/Drasil/Expr/Lang.hs

```diff
@@ -232,82 +238,93 @@ instance LiteralC Expr where
 
 
 -- helper function for typechecking to help reduce duplication
-vvvInfer :: TypingContext Space -> VVVBinOp -> Expr -> Expr -> Either Space TypeError
-vvvInfer ctx op l r = case (infer ctx l, infer ctx r) of
-    (Left lt@(S.Vect lsp), Left (S.Vect rsp)) ->
-      if lsp == rsp && S.isBasicNumSpace lsp then
-        if op == VSub && (lsp == S.Natural || rsp == S.Natural) then
-          Right $ "Vector subtraction expects both operands to be vectors of non-natural numbers. Received `" ++ show lsp ++ "` and `" ++ show rsp ++ "`."
-        else Left lt
-      else Right $ "Vector " ++ pretty op ++ " expects both operands to be vectors of non-natural numbers. Received `" ++ show lsp ++ "` and `" ++ show rsp ++ "`."
-    (Left lsp, Left rsp) -> Right $ "Vector operation " ++ pretty op ++ " expects vector operands. Received `" ++ show lsp ++ "` and `" ++ show rsp ++ "`."
-    (_       , Right re) -> Right re
-    (Right le, _       ) -> Right le
+vvvInfer :: TypingContext Space -> VVVBinOp -> Expr -> Expr -> Either TypeError Space
+vvvInfer ctx op l r = do
+  lt <- infer ctx l
+  rt <- infer ctx r
+
+  let msg dir sp = "Vector operation " ++ pretty op ++ " expects numeric vectors, but found `" ++ sp ++ "` on the " ++ dir ++ "-hand side."
+
+  lsp <- assertNumericVector lt $ msg "left"
+  rsp <- assertNumericVector rt $ msg "right"
+
+  if op == VSub then
+    assertNonNatNumeric lsp $ \sp ->
+      "Vector subtraction expects both operands to be vectors of non-natural numbers. Received `" ++ sp ++ "`."
+  else Right ()
 
+  lsp ~== rsp $ \lt' rt' -> "Vector " ++ pretty op ++ " expects both operands to be of the same numeric type. Received `" ++ lt' ++ "` and `" ++ rt' ++ "`."
+
+  pure lt
 
 instance Typed Expr Space where
-  check :: TypingContext Space -> Expr -> Space -> Either Space TypeError
+  check :: TypingContext Space -> Expr -> Space -> Either TypeError Space
   check = typeCheckByInfer
 
-  infer :: TypingContext Space -> Expr -> Either Space TypeError
+  infer :: TypingContext Space -> Expr -> Either TypeError Space
   infer cxt (Lit lit) = infer cxt lit
 
-  infer cxt (AssocA _ (e:exs)) = 
+  infer cxt (AssocA _ (e:exs)) = do
+    et <- infer cxt e
+    assertNumeric et $
+      \sp -> "Associative arithmetic operation expects numeric operands, but found `" ++ sp ++ "`."
+    assertAllEq cxt exs et
+        "Associative arithmetic operation expects all operands to be of the same type."
+    pure et
+  infer _ (AssocA Add _) = Left "Associative addition requires at least one operand."
```

Comment:
Actually, a 0-ary addition is '0', so this could be legal. Biggest issue is that we can't `infer` such a thing, we can only `check` it. Which makes me think: should `AssocA` really be an `infer` type?

Reply:
Yes, we should probably change this to a `check`, and I think we need to review all the type checking rules to really make use of a bidirectional type checker. We only `infer` right now, really.

Reply:
Subsequent work, I assume?

Reply:
Yes. So I'll remove my 'changes requested'.